lib/galaxy/jobs/runners/arc.py

```diff
@@ -0,0 +1,635 @@
+        return remote_extra_files_path_rewrite
+
+    def input_metadata_rewrite(self, dataset, metadata_val):
+        """TODO - find out what this is and if I need it"""
+        return None
+
+    def unstructured_path_rewrite(self, parameter_value):
+        """TODO - find out what this is and if I need it"""
+        return self._working_directory
+
+    def working_directory(self):
+        return self._working_directory
+
+    def env_config_directory(self):
+        return self._config_directory
+
+    def config_directory(self):
+        return self._config_directory
+
+    def new_file_path(self):
+        return self._working_directory
+
+    def sep(self):
+        return self._sep
+
+    def version_path(self):
+        return self._version_path
+
+    def tool_directory(self):
+        return self._tool_dir
+
+    def home_directory(self):
+        return self._home_directory
+
+    def tmp_directory(self):
+        return self._tmp_directory
+
+    def galaxy_url(self):
+        return self.job_wrapper.get_destination_configuration("galaxy_infrastructure_url")
+
+    def get_file_sources_dict(self):
+        return self.job_wrapper.job_io.file_sources_dict
+
+
+class ArcRESTJobRunner(AsynchronousJobRunner):
+    """
+    Job runner backed by a finite pool of worker threads. FIFO scheduling
+    """
+
+    runner_name = "ArcRESTJobRunner"
+
+    def __init__(self, app, nworkers, **kwargs):
+        """1: Get runner_param_specs from job_conf.xml
+        2: Initialise job runner parent object
+        3: Start the worker and monitor threads
+        """
+
+        # Start the job runner parent object
+        super().__init__(app, nworkers, **kwargs)
+        ensure_pyarc()
+
+        self.arc = None
+        self.arcjob = None
+        self.provider_backend = provider_name_to_backend("wlcg")
+
+    def queue_job(self, job_wrapper):
+        """When a tool is submitted for execution in galaxy"""
+        """ This method
+        1. Fetches the configured ARC endpoint for this user
+        2. Prepares an ARC job description based on the jobs destination parameters
+        3. Submits the job to the remote ARC endpoint via pyarcrest
+        4. Adds the job to the galaxy job queue
+        """
+
+        job_destination = job_wrapper.job_destination
+        job_id = job_wrapper.job_id
+
+        """ Build the command line - needs a rewrite of input-paths as ARC is a remote cluster """
+        if not self.prepare_job(
+            job_wrapper,
+            include_metadata=False,
+            include_work_dir_outputs=False,
+            modify_command_for_container=False,
+            stream_stdout_stderr=False,
+        ):
+            return
+        """ prepare_job() calls prepare() but not allowing to pass a compute_environment object
+        As I need to define my own compute_environment for the remote compute I must call it here passing the compute_environment
+        TODO - not a good solution"""
```

Comment:
I'm facing a similar problem with the DIRAC jobrunner and thought of modifying the `prepare_job` to take an extra argument `compute_environment` (default `None`), so it should keep working with existing implementations but allow for customisation